### ISSUE TITLE
module memory tracking (Trac #722)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#716]
I'd like to see memory usage tracking at the module level, similar to cpu usage tracking that we do now.  Below are the notes and links I gathered a while ago, placed here so I don't lose them.

- modify allocator to track allocations/deallocations by the module that made it
    - simplistic approach is to tag each allocation with the current tag
    - keep stack of tags that can be pushed to / popped from
    - push tag on module constructor, Configure_, Process_, Finish (any others?)

LD_PRELOAD examples:
https://github.com/jtolds/malloc_instrumentation/blob/master/malloc_instrument.c
http://blog.oakbits.com/index.php?post/2012/04/11/LD_PRELOAD-for-real-usage-on-memory-allocation

GNU malloc hook docs and example:
http://www.gnu.org/software/libc/manual/html_node/Hooks-for-Malloc.html
http://code.google.com/p/jmem/source/browse/trunk/include/jmem.h

Tracking by file and line in c++:
http://www.almostinfinite.com/memtrack.html

Pre-processor redefining of malloc:
http://cboard.cprogramming.com/c-programming/110860-tracking-memory-malloc-free.html

Straight overriding of malloc:
http://panthema.net/2013/malloc_count/

Some interesting notes on python memory tracking:
http://legacy.python.org/dev/peps/pep-0454/ 

Migrated from https://code.icecube.wisc.edu/ticket/722
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:12:58", 
    "description": "I'd like to see memory usage tracking at the module level, similar to cpu usage tracking that we do now.  Below are the notes and links I gathered a while ago, placed here so I don't lose them.\n\n\n- modify allocator to track allocations/deallocations by the module that made it\n    - simplistic approach is to tag each allocation with the current tag\n    - keep stack of tags that can be pushed to / popped from\n    - push tag on module constructor, Configure_, Process_, Finish (any others?)\n\nLD_PRELOAD examples:\nhttps://github.com/jtolds/malloc_instrumentation/blob/master/malloc_instrument.c\nhttp://blog.oakbits.com/index.php?post/2012/04/11/LD_PRELOAD-for-real-usage-on-memory-allocation\n\nGNU malloc hook docs and example:\nhttp://www.gnu.org/software/libc/manual/html_node/Hooks-for-Malloc.html\nhttp://code.google.com/p/jmem/source/browse/trunk/include/jmem.h\n\nTracking by file and line in c++:\nhttp://www.almostinfinite.com/memtrack.html\n\nPre-processor redefining of malloc:\nhttp://cboard.cprogramming.com/c-programming/110860-tracking-memory-malloc-free.html\n\nStraight overriding of malloc:\nhttp://panthema.net/2013/malloc_count/\n\nSome interesting notes on python memory tracking:\nhttp://legacy.python.org/dev/peps/pep-0454/ \n", 
    "reporter": "david.schultz", 
    "cc": "nega, olivas, cweaver", 
    "resolution": "fixed", 
    "_ts": "1550067178841456", 
    "component": "combo core", 
    "summary": "module memory tracking", 
    "priority": "minor", 
    "keywords": "", 
    "time": "2014-03-21T15:58:26", 
    "milestone": "", 
    "owner": "david.schultz", 
    "type": "enhancement"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20150323T22:50:10: 
* olivas changed component from "IceTray" to "icecube offline"
* olivas commented:

> Another IceTray -> icecube offline edit.

**Comment by icecube-trac** [comment]
Trac update at 20150810T22:59:36: olivas changed milestone from "" to "January 2016"

**Comment by icecube-trac** [comment]
Trac update at 20151013T13:58:39: 
* david.schultz changed milestone from "January 2016" to "offline V15-11-00"
* david.schultz changed owner from "nega" to "david.schultz"
* david.schultz changed status from "new" to "accepted"
* david.schultz commented:

> First attempt shall try memtrack.  Needs adding to the uber header in cmake.  Will make a branch and update here in the next few days.

**Comment by icecube-trac** [comment]
Trac update at 20151014T12:32:01: david.schultz commented:

> Did a simple MemTrack test here:
> source:IceCube/sandbox/dschultz/memtrack
> 
> It allows instantaneous knowledge of all heap memory, but doesn't track maximums for printing at exit.  Maybe that's good enough?  Sending SIGHUP when you want the info is the easiest way to use this.

**Comment by icecube-trac** [comment]
Trac update at 20151102T21:25:34: david.schultz commented:

> Lots of the pybindings give this error:
> 
> ```text
> *** Error in `python': munmap_chunk(): invalid pointer: 0x00000000027fe408 ***
> ```
> 
> Not sure exactly what's up, except tracked memory going across the border with python and not being freed properly.
> 
> I'll try the LD_PRELOAD method and see if it has the same problems.

**Comment by icecube-trac** [comment]
Trac update at 20151102T23:43:00: david.schultz commented:

> LD_PRELOAD with malloc_instrument.c works on regular c code, but segfaults on c++ for some reason. Trying another one.

**Comment by icecube-trac** [comment]
Trac update at 20151104T20:04:31: 
* david.schultz changed _comment0 from:

> malloc_count has a test here:
> source:IceCube/sandbox/dschultz/memtrack
> 
> Works fine if you're doing `malloc()`, but doesn't see `new`. This is problematic, since we like c++.

to:

> 1446751327674076


* david.schultz commented:

> malloc_count has a test here:
> source:IceCube/sandbox/dschultz/memtrack3
> 
> Works fine if you're doing `malloc()`, but doesn't see `new`. This is problematic, since we like c++.

**Comment by icecube-trac** [comment]
Trac update at 20151104T20:10:20: david.schultz commented:

> This external tool would be somewhat promising for getting a general overview:
> https://github.com/KDE/heaptrack
> 
> But, installing it is a pain (still haven't figured out all the dependencies).

**Comment by icecube-trac** [comment]
Trac update at 20151104T22:03:20: david.schultz commented:

> Some approaches:
> 
> 1. Use an external tool (heaptrack, massif)
> 
> ```text
> + easier to implement
> 
> - only works on the dev machine where it is installed
> - doesn't give us module-level usage
> ```
> 
> 2. Play around some more with LD_PRELOAD and C++
> 
> ```text
> + should catch things from python land
> 
> - doesn't give us module-level usage
> ```
> 
> 3. Work on MemTrack's python memory passing issue
> 
> ```text
> + plenty of flexibility
> + can get module-level usage
> 
> - no memory tracking for things from python
> ```
> 
> 4. Use a special allocator for icecube software
> 
> ```text
> + full flexibility
> 
> - no memory tracking for things from python
> - lots of invasive work
> ```

**Comment by icecube-trac** [comment]
Trac update at 20151105T19:54:45: david.schultz commented:

> Apparently python has built-in support for memory tracing since 3.4:
> https://docs.python.org/3.5/library/tracemalloc.html
> 
> The output is vastly superior to what you can see from the C side.  So, I'd say to not worry about the python memory tracing yet, and handle it natively when we need to.

**Comment by icecube-trac** [comment]
Trac update at 20151201T16:22:03: david.schultz changed cc from "" to "nega, olivas, cweaver"

**Comment by icecube-trac** [comment]
Trac update at 20160120T22:52:55: david.schultz changed milestone from "offline V15-12-00" to "March 2016"

**Comment by icecube-trac** [comment]
Trac update at 20160215T16:55:33: david.schultz commented:

> I've played with MemTrack a bit more, and probably the biggest hurdle is that the `boost::type_traits` header doesn't play nice with it (there is an `#error` for defining `new` as a macro). It works fine if you include boost before MemTrack, so the easy solution is to get the header order fixed in all the files.
> 
> Another (possibly better) solution is to not do a global replace of the `new` operator, but instead make a custom new that we call only inside our code. This means we can't track any external allocations, but that's probably fine.
> 
> Either way is fairly invasive, so probably needs committee approval.

**Comment by icecube-trac** [comment]
Trac update at 20160219T22:21:40: david.schultz commented:

> From slack, by cweaver:
> 
> dschultz: No time to write anything new (no pun intended).
> However, Here's a really bare-bones version I use in the SQuIDS tests: https://github.com/jsalvado/SQuIDS/blob/master/test/alloc_counting.h
> 
> It does nothing except count bytes allocated and optionally fill fresh memory with a noticeable pattern.
> 
> In principle one can do much more, but the trick is to do it without triggering recursive memory allocations that blow up.
> 
> I.e. it would be a bad idea for operator new to internally use high level data structures which themselves call new.
>

**Comment by icecube-trac** [comment]
Trac update at 20160219T22:24:26: david.schultz commented:

> Playing with this a bit more, part of the cleverness using `#define new` gives you is the ability to get the file, line number, and type information.  That's kind of valuable.  Otherwise, we just know we allocated N bytes.
> 
> Granted, we still might have enough information with just the number of bytes and a helper inside icetray to tell us which module is currently running.

**Comment by icecube-trac** [comment]
Trac update at 20160219T22:27:38: cweaver commented:

> "gives you is the ability to get the file, line number, and type information" This sort of thing is why I was thinking of trying to use libunwind. Then you could in principle get the whole call stack, as well, although for practical reasons I would probably just store the information about the last icetray function seen on the stack.

**Comment by icecube-trac** [comment]
Trac update at 20160301T16:10:40: david.schultz commented:

> A branch is available, if you haven't already seen it:
> source:IceTray/projects/icetray/branches/memory_tracking
> 
> It seems to work well on a small test program.

**Comment by icecube-trac** [comment]
Trac update at 20160420T18:52:16: olivas changed milestone from "March 2016" to "April 2016"

**Comment by icecube-trac** [comment]
Trac update at 20160616T19:23:04: david.schultz commented:

> As I said at my presentation in Stony Brook, the branch is stable and ready to be merged.  So, I'll do that soon unless there are objections.

**Comment by icecube-trac** [comment]
Trac update at 20160617T14:49:28: 
* david.schultz changed status from "accepted" to "closed"
* david.schultz commented:

> In changeset 2605:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="2605"
> merge memory_tracking branch. closes #716
> ```


* david.schultz changed resolution from "" to "fixed"

**Comment by icecube-trac** [comment]
Trac update at 20190111T20:49:13: nwhitehorn commented:

> In changeset 168752:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="168752"
> merge memory_tracking branch. closes #716
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:12:58: 
* olivas changed milestone from "May 2016" to ""
* olivas commented:

> Milestone May 2016 deleted
